### PR TITLE
Task-41305: Add last connection Column in userManagement

### PR DESCRIPTION
**Pull request by AzmiTouil**

### CODE REVIEW

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java [hunk 1]
```diff
@@ -188,13 +189,22 @@ public static ProfileEntity buildEntityProfile(Profile profile, String restPath,
...
       OrganizationService organizationService = getOrganizationService();
       User user = organizationService.getUserHandler().findUserByName(userEntity.getUsername());
       if (user != null) {
         userEntity.setIsInternal(user.isInternalStore());
+        if (isExternal && user.getLastLoginTime().equals(user.getCreatedDate())) {
+          userEntity.setLastConnexion("neverConnected");
+        } else if (user.getLastLoginTime().equals(user.getCreatedDate())) {
+          userEntity.setLastConnexion("invitedToJoin");
+        } else {
+          String lastLogin = TimeConvertUtils.getFormatDate(user.getLastLoginTime(), "dd/MM/yyyy HH:mm");
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: I think it's better to use this reusable component to be able to display date with adequate Date Format switch user language:
https://github.com/Meeds-io/social/blob/develop/webapp/portlet/src/main/webapp/common/components/DateFormat.vue
@azayati

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java [hunk 2]
```diff
@@ -188,13 +188,21 @@ public static ProfileEntity buildEntityProfile(Profile profile, String restPath,
     userEntity.setAvatar(profile.getAvatarUrl());
     userEntity.setBanner(profile.getBannerUrl());
     try {
+      boolean isExternal = profile.getProperty(Profile.EXTERNAL) != null && (profile.getProperty(Profile.EXTERNAL)).equals("true");
```
Comment (rdenarie) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: Do we already have external information in profile ? I dont remember that we put it

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java [hunk 3]
```diff
@@ -188,13 +188,21 @@ public static ProfileEntity buildEntityProfile(Profile profile, String restPath,
...
       User user = organizationService.getUserHandler().findUserByName(userEntity.getUsername());
       if (user != null) {
         userEntity.setIsInternal(user.isInternalStore());
+        if (isExternal && user.getLastLoginTime().equals(user.getCreatedDate())) {
+          userEntity.setLastConnexion("neverConnected");
+        } else if (user.getLastLoginTime().equals(user.getCreatedDate())) {
+          userEntity.setLastConnexion("neverEnrolled");
+        } else {
+          userEntity.setLastConnexion(String.valueOf(user.getLastLoginTime().getTime()));
+        }
```
Comment (rdenarie) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: we should not have 3 states here, only 2 : the date if exist, null else. 
And at display, we will check the enrollDate (that we dont have for the moment)

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java (hunk 2 above)
Comment (AzmiTouil) [reply component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: yes , we have already stored the info 
https://github.com/Meeds-io/social/blob/786db1a3f5d59d7e4050c58067cd77b2678127df/component/core/src/main/java/org/exoplatform/social/core/listeners/SocialMembershipListenerImpl.java#L146

Comment (rdenarie) [reply component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: :+1:

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java [hunk 4]
```diff
@@ -192,9 +192,14 @@ public static ProfileEntity buildEntityProfile(Profile profile, String restPath,
       User user = organizationService.getUserHandler().findUserByName(userEntity.getUsername());
       if (user != null) {
         userEntity.setIsInternal(user.isInternalStore());
+        if (user.getLastLoginTime().equals(user.getCreatedDate())) {
+          userEntity.setLastConnexion(null);
+        } else {
+          userEntity.setLastConnexion(String.valueOf(user.getLastLoginTime().getTime()));
+        }
```
Comment (rdenarie) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]:
```suggestion
        if (!user.getLastLoginTime().equals(user.getCreatedDate())) {
                userEntity.setLastConnexion(String.valueOf(user.getLastLoginTime().getTime()));
        }
```
Is it sufficient ?

Comment (AzmiTouil) [reply component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java:199]: For this PR , we will don't treat the enrollment date , we need just to compare the last connection date with the creation date to know if the user has logged or not
so yes it sufficient.

File: webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementList.vue [hunk 5]
```diff
@@ -29,6 +29,21 @@
...
+      <template slot="item.lastConnexion" slot-scope="{ item }">
+        <div v-if="typeof item.lastConnexion == 'number'">
+          <date-format
+            :value="item.lastConnexion"
+            :format="fullDateFormat"
+            class="mr-1" />
+          <date-format
+            :value="item.lastConnexion"
+            :format="dateTimeFormat"
+            class="mr-1" />
```
Comment (boubaker) [inline webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementList.vue]:
```suggestion
          <date-format
            :value="item.lastConnexion"
            :format="fullDateFormat"
            class="mr-1" />
```
You can combine both by using this object:

```
    fullDateFormat: {
      day: 'numeric',
      month: 'short',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    }
```

File: webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementList.vue [hunk 6]
```diff
@@ -80,6 +95,15 @@ export default {
...
     error: null,
+    fullDateFormat: {
+      day: 'numeric',
+      month: 'short',
+      year: 'numeric',
+    },
+    dateTimeFormat: {
+      hour: '2-digit',
+      minute: '2-digit',
+    }
```
Comment (boubaker) [inline webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementList.vue]: You can combine both

```suggestion
    fullDateFormat: {
      day: 'numeric',
      month: 'short',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    }
```

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java (hunk 4 above)
Comment (AzmiTouil) [reply component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: yes, updated!

File: webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementList.vue (hunk 6 above)
Comment (AzmiTouil) [reply webapp/portlet/src/main/webapp/idm-users-management/components/UsersManagementList.vue]: Updated! thanks @boubaker

File: component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java [hunk 7]
```diff
@@ -192,9 +192,12 @@ public static ProfileEntity buildEntityProfile(Profile profile, String restPath,
       User user = organizationService.getUserHandler().findUserByName(userEntity.getUsername());
       if (user != null) {
         userEntity.setIsInternal(user.isInternalStore());
+        if (!user.getLastLoginTime().equals(user.getCreatedDate())) {
+          userEntity.setLastConnexion(String.valueOf(user.getLastLoginTime().getTime()));
+        }
       }
     } catch (Exception e) {
-      LOG.warn("Error while checking internal store user", e);
+      LOG.warn("Error when searching user " + userEntity.getUsername(), e);
```
Comment (boubaker) [inline component/service/src/main/java/org/exoplatform/social/rest/api/EntityBuilder.java]: Use Message template better than string concatenation in logs

```suggestion
      LOG.warn("Error when searching user {}", userEntity.getUsername(), e);
```

**Review by rdenarie — CHANGES_REQUESTED** [review]
Just fix the Boubaker's comment and it is ok.

**Review by rdenarie — DISMISSED** [review]

**Review by rdenarie — DISMISSED** [review]

**Review by rdenarie — APPROVED** [review]

**Pull request merged**